File: data/samples/benign/jira-transitions/scripts/workflow-demo.py
#!/usr/bin/env python3
"""
Jira Workflow Demo (Python)
Demonstrates workflow transitions: To Do -> Progressing -> Done
Following jira-safe skill patterns for Next-Gen projects.

Usage:
  python jira-workflow-demo.py demo SCRUM-123
  python jira-workflow-demo.py start SCRUM-123
  python jira-workflow-demo.py complete SCRUM-123
  python jira-workflow-demo.py reopen SCRUM-123
  python jira-workflow-demo.py status SCRUM-123
"""

import base64
import json
import os
import sys
import time
from pathlib import Path
from urllib.request import Request, urlopen
from urllib.error import HTTPError

# Load .env file from jira root (two levels up from scripts/)
def load_env():
    env_path = Path(__file__).parent.parent.parent / '.env'
    if env_path.exists():
        with open(env_path, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#') and '=' in line:
                    key, value = line.split('=', 1)
                    os.environ.setdefault(key.strip(), value.strip())

load_env()

# Configuration from environment variables
JIRA_EMAIL = os.environ.get('JIRA_EMAIL')
JIRA_API_TOKEN = os.environ.get('JIRA_API_TOKEN')
JIRA_BASE_URL = os.environ.get('JIRA_BASE_URL')
PROJECT_KEY = os.environ.get('JIRA_PROJECT_KEY', 'SCRUM')

# Validate required env vars
if not all([JIRA_EMAIL, JIRA_API_TOKEN, JIRA_BASE_URL]):
    print('Error: Missing required environment variables.', file=sys.stderr)
    print('Required: JIRA_EMAIL, JIRA_API_TOKEN, JIRA_BASE_URL', file=sys.stderr)
    print('Set these in .claude/skills/jira/.env or export them manually.', file=sys.stderr)
    sys.exit(1)

# Build auth header
auth_string = f'{JIRA_EMAIL}:{JIRA_API_TOKEN}'
auth_bytes = base64.b64encode(auth_string.encode('utf-8')).decode('utf-8')

HEADERS = {
    'Authorization': f'Basic {auth_bytes}',
    'Content-Type': 'application/json',
    'Accept': 'application/json'
}


def make_request(method, path, data=None):
    """Make HTTP request to Jira API."""
    url = f'{JIRA_BASE_URL}/rest/api/3{path}'

    body = json.dumps(data).encode('utf-8') if data else None
    req = Request(url, data=body, headers=HEADERS, method=method)

    try:
        with urlopen(req) as response:
            status = response.status
            if status == 204:
                return None
            return json.loads(response.read().decode('utf-8'))
    except HTTPError as e:
        error_body = e.read().decode('utf-8')
        raise Exception(f'{e.code}: {error_body[:200]}')


def get_issue(issue_key):
    """Get issue details."""
    return make_request('GET', f'/issue/{issue_key}?fields=summary,status')


def get_transitions(issue_key):
    """Get available transitions for an issue."""
    result = make_request('GET', f'/issue/{issue_key}/transitions?expand=transitions.fields')
    return result['transitions']


def do_transition(issue_key, transition_id):
    """Execute a transition."""
    data = {'transition': {'id': transition_id}}
    make_request('POST', f'/issue/{issue_key}/transitions', data)


def add_comment(issue_key, comment_text):
    """Add a comment to an issue."""
    data = {
        'body': {
            'type': 'doc',
            'version': 1,
            'content': [
                {
                    'type': 'paragraph',
                    'content': [{'type': 'text', 'text': comment_text}]
                }
            ]
        }
    }
    make_request('POST', f'/issue/{issue_key}/comment', data)


def transition_to(issue_key, target_status):
    """Transition issue to a target status."""
    transitions = get_transitions(issue_key)

    # Find matching transition
    transition = None
    for t in transitions:
        if t['to']['name'].lower() == target_status.lower() or t['name'].lower() == target_status.lower():
            transition = t
            break

    if not transition:
        print(f'No transition to "{target_status}" available.')
        print('Available transitions:')
        for t in transitions:
            print(f'  - {t["name"]} -> {t["to"]["name"]}')
        return False

    do_transition(issue_key, transition['id'])
    return True


def show_status(issue_key):
    """Show current issue status."""
    issue = get_issue(issue_key)
    status = issue['fields']['status']['name']
    summary = issue['fields']['summary']
    print(f'{issue_key}: {status}')
    print(f'Summary: {summary}')


def run_demo(issue_key):
    """Run full workflow demo."""
    print('=' * 40)
    print('  JIRA WORKFLOW DEMO (PYTHON)')
    print('=' * 40)
    print(f'\nIssue: {issue_key}')
    print(f'Project: {PROJECT_KEY}')
    print(f'URL: {JIRA_BASE_URL}/browse/{issue_key}')

    # Step 1: Check current status
    print('\n--- Step 1: Check Current Status ---')
    issue = get_issue(issue_key)
    current_status = issue['fields']['status']['name']
    print(f'Initial status: {current_status}')

    # Step 2: Show available transitions
    print('\n--- Step 2: Available Transitions ---')
    transitions = get_transitions(issue_key)
    print(f'Found {len(transitions)} transitions:')
    for t in transitions:
        print(f'  - {t["id"]}: "{t["name"]}" -> {t["to"]["name"]}')

    # Step 3: Start work (transition to Progressing)
    print('\n--- Step 3: Start Work ---')
    print(f'\nStarting work on {issue_key}...')
    issue = get_issue(issue_key)
    print(f'  Current status: {issue["fields"]["status"]["name"]}')

    if transition_to(issue_key, 'Progressing'):
        add_comment(issue_key, '[Python Demo] Work started')
        print('  Transitioned to: Progressing')
        print('  Added start comment')

    # Step 4: Simulate work
    print('\n--- Step 4: Simulating Work ---')
    print('  [Simulating development work...]')
    time.sleep(1)
    print('  [Work complete!]')

    # Step 5: Complete work
    print('\n--- Step 5: Complete Work ---')
    print(f'\nCompleting work on {issue_key}...')
    issue = get_issue(issue_key)
    print(f'  Current status: {issue["fields"]["status"]["name"]}')

    if transition_to(issue_key, 'Done'):
        add_comment(issue_key, '[Python Demo] Work completed')
        print('  Transitioned to: Done')
        print('  Added completion comment')

    # Step 6: Final status
    print('\n--- Step 6: Final Status ---')
    issue = get_issue(issue_key)
    print(f'Final status: {issue["fields"]["status"]["name"]}')

    print('\n' + '=' * 40)
    print('  WORKFLOW DEMO COMPLETE')
    print('=' * 40)
    print(f'\nView issue: {JIRA_BASE_URL}/browse/{issue_key}')


def main():
    if len(sys.argv) < 2:
        print('Usage:')
        print('  python jira-workflow-demo.py demo SCRUM-55')
        print('  python jira-workflow-demo.py start SCRUM-55')
        print('  python jira-workflow-demo.py complete SCRUM-55')
        print('  python jira-workflow-demo.py reopen SCRUM-55')
        print('  python jira-workflow-demo.py status SCRUM-55')
        sys.exit(0)

    action = sys.argv[1]
    issue_key = sys.argv[2] if len(sys.argv) > 2 else None

    if not issue_key and action != 'help':
        print('Error: Issue key required')
        sys.exit(1)

    try:
        if action == 'demo':
            run_demo(issue_key)
        elif action == 'start':
            print(f'Starting work on {issue_key}...')
            if transition_to(issue_key, 'Progressing'):
                add_comment(issue_key, 'Work started')
                print(f'[OK] {issue_key} transitioned to Progressing')
        elif action == 'complete':
            print(f'Completing work on {issue_key}...')
            if transition_to(issue_key, 'Done'):
                add_comment(issue_key, 'Work completed')
                print(f'[OK] {issue_key} transitioned to Done')
        elif action == 'reopen':
            print(f'Reopening {issue_key}...')
            if transition_to(issue_key, 'To Do'):
                add_comment(issue_key, 'Issue reopened')
                print(f'[OK] {issue_key} transitioned to To Do')
        elif action == 'status':
            show_status(issue_key)
        else:
            print(f'Unknown action: {action}')
            sys.exit(1)

    except Exception as e:
        print(f'Error: {e}')
        sys.exit(1)


if __name__ == '__main__':
    main()

File: data/samples/benign/jira-transitions/scripts/workflow-demo.mjs
// Jira Workflow Demo: Backlog -> In Progress -> Done
// Demonstrates the complete workflow cycle

// Load from environment variables (set by run.js or manually)
const JIRA_EMAIL = process.env.JIRA_EMAIL;
const JIRA_API_TOKEN = process.env.JIRA_API_TOKEN;
const JIRA_BASE_URL = process.env.JIRA_BASE_URL;
const PROJECT_KEY = process.env.JIRA_PROJECT_KEY || 'SCRUM';

// Validate required env vars
if (!JIRA_EMAIL || !JIRA_API_TOKEN || !JIRA_BASE_URL) {
  console.error('Error: Missing required environment variables.');
  console.error('Required: JIRA_EMAIL, JIRA_API_TOKEN, JIRA_BASE_URL');
  console.error('Set these in .claude/skills/jira/.env or export them manually.');
  process.exit(1);
}

const auth = Buffer.from(`${JIRA_EMAIL}:${JIRA_API_TOKEN}`).toString('base64');
const headers = {
  'Authorization': `Basic ${auth}`,
  'Content-Type': 'application/json',
  'Accept': 'application/json',
};

// ==================== API HELPERS ====================

async function jiraRequest(path, options = {}) {
  const url = `${JIRA_BASE_URL}/rest/api/3${path}`;
  const response = await fetch(url, { ...options, headers });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`${response.status}: ${error.substring(0, 200)}`);
  }

  if (response.status === 204) return null;
  return response.json();
}

async function getIssueStatus(issueKey) {
  const issue = await jiraRequest(`/issue/${issueKey}?fields=status`);
  return issue.fields.status.name;
}

async function getTransitions(issueKey) {
  const response = await jiraRequest(`/issue/${issueKey}/transitions`);
  return response.transitions;
}

async function transitionTo(issueKey, targetState) {
  const transitions = await getTransitions(issueKey);

  const transition = transitions.find(t =>
    t.to.name.toLowerCase() === targetState.toLowerCase() ||
    t.name.toLowerCase() === targetState.toLowerCase()
  );

  if (!transition) {
    console.log(`Available transitions for ${issueKey}:`);
    transitions.forEach(t => console.log(`  - ${t.id}: ${t.name} → ${t.to.name}`));
    throw new Error(`No transition to "${targetState}" found`);
  }

  await jiraRequest(`/issue/${issueKey}/transitions`, {
    method: 'POST',
    body: JSON.stringify({ transition: { id: transition.id } })
  });

  return transition.to.name;
}

async function addComment(issueKey, text) {
  const body = {
    body: {
      type: 'doc',
      version: 1,
      content: [{ type: 'paragraph', content: [{ type: 'text', text }] }]
    }
  };

  await jiraRequest(`/issue/${issueKey}/comment`, {
    method: 'POST',
    body: JSON.stringify(body)
  });
}

// ==================== WORKFLOW OPERATIONS ====================

async function startWork(issueKey) {
  console.log(`\nStarting work on ${issueKey}...`);

  const beforeStatus = await getIssueStatus(issueKey);
  console.log(`  Current status: ${beforeStatus}`);

  if (beforeStatus.toLowerCase() === 'progressing') {
    console.log(`  Already progressing, skipping transition`);
    return beforeStatus;
  }

  const newStatus = await transitionTo(issueKey, 'Progressing');
  console.log(`  Transitioned to: ${newStatus}`);

  await addComment(issueKey, `Work started at ${new Date().toISOString()}`);
  console.log(`  Added start comment`);

  return newStatus;
}

async function completeWork(issueKey, details = {}) {
  console.log(`\nCompleting work on ${issueKey}...`);

  const beforeStatus = await getIssueStatus(issueKey);
  console.log(`  Current status: ${beforeStatus}`);

  if (beforeStatus.toLowerCase() === 'done') {
    console.log(`  Already done, skipping transition`);
    return beforeStatus;
  }

  const newStatus = await transitionTo(issueKey, 'Done');
  console.log(`  Transitioned to: ${newStatus}`);

  // Add completion comment
  let commentText = `Work completed at ${new Date().toISOString()}`;
  if (details.files) {
    commentText += `\n\nFiles modified:\n${details.files.map(f => `- ${f}`).join('\n')}`;
  }
  if (details.commits) {
    commentText += `\n\nCommits:\n${details.commits.map(c => `- ${c}`).join('\n')}`;
  }
  if (details.notes) {
    commentText += `\n\nNotes: ${details.notes}`;
  }

  await addComment(issueKey, commentText);
  console.log(`  Added completion comment`);

  return newStatus;
}

async function reopenWork(issueKey) {
  console.log(`\nReopening ${issueKey}...`);

  const beforeStatus = await getIssueStatus(issueKey);
  console.log(`  Current status: ${beforeStatus}`);

  const newStatus = await transitionTo(issueKey, 'To Do');
  console.log(`  Transitioned to: ${newStatus}`);

  await addComment(issueKey, `Reopened at ${new Date().toISOString()}`);

  return newStatus;
}

// ==================== DEMO WORKFLOW ====================

async function demoWorkflow(issueKey) {
  console.log('========================================');
  console.log('  JIRA WORKFLOW DEMO');
  console.log('========================================\n');
  console.log(`Issue: ${issueKey}`);
  console.log(`Project: ${PROJECT_KEY}`);
  console.log(`URL: ${JIRA_BASE_URL}/browse/${issueKey}\n`);

  // 1. Check current status
  console.log('--- Step 1: Check Current Status ---');
  const initialStatus = await getIssueStatus(issueKey);
  console.log(`Initial status: ${initialStatus}\n`);

  // 2. Show available transitions
  console.log('--- Step 2: Available Transitions ---');
  const transitions = await getTransitions(issueKey);
  console.log(`Found ${transitions.length} transitions:`);
  for (const t of transitions) {
    console.log(`  - ${t.id}: "${t.name}" → ${t.to.name}`);
  }
  console.log('');

  // 3. Start work (move to In Progress)
  console.log('--- Step 3: Start Work ---');
  await startWork(issueKey);

  // 4. Simulate doing work
  console.log('\n--- Step 4: Simulating Work ---');
  console.log('  [Simulating development work...]');
  await new Promise(r => setTimeout(r, 1000));
  console.log('  [Work complete!]\n');

  // 5. Complete work (move to Done)
  console.log('--- Step 5: Complete Work ---');
  await completeWork(issueKey, {
    files: ['app/api/example/route.ts', 'lib/utils/helper.ts'],
    commits: ['abc1234 - feat: implement feature', 'def5678 - test: add tests'],
    notes: 'Demo workflow completed successfully'
  });

  // 6. Final status
  console.log('\n--- Step 6: Final Status ---');
  const finalStatus = await getIssueStatus(issueKey);
  console.log(`Final status: ${finalStatus}`);

  console.log('\n========================================');
  console.log('  WORKFLOW DEMO COMPLETE');
  console.log('========================================');
  console.log(`\nView issue: ${JIRA_BASE_URL}/browse/${issueKey}`);

  return { initialStatus, finalStatus };
}

// ==================== BATCH WORKFLOW ====================

async function batchWorkflow(issueKeys, operation) {
  console.log('========================================');
  console.log(`  BATCH ${operation.toUpperCase()}`);
  console.log('========================================\n');
  console.log(`Processing ${issueKeys.length} issues...\n`);

  const results = { success: [], failed: [] };

  for (const key of issueKeys) {
    try {
      if (operation === 'start') {
        await startWork(key);
      } else if (operation === 'complete') {
        await completeWork(key);
      } else if (operation === 'reopen') {
        await reopenWork(key);
      }
      results.success.push(key);
    } catch (error) {
      console.log(`  FAILED ${key}: ${error.message}`);
      results.failed.push({ key, error: error.message });
    }

    await new Promise(r => setTimeout(r, 200)); // Rate limiting
  }

  console.log('\n========================================');
  console.log('  BATCH COMPLETE');
  console.log('========================================');
  console.log(`Success: ${results.success.length}`);
  console.log(`Failed: ${results.failed.length}`);

  return results;
}

// ==================== MAIN ====================

async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log('Usage:');
    console.log('  node jira-workflow-demo.mjs demo SCRUM-55');
    console.log('  node jira-workflow-demo.mjs start SCRUM-55');
    console.log('  node jira-workflow-demo.mjs complete SCRUM-55');
    console.log('  node jira-workflow-demo.mjs reopen SCRUM-55');
    console.log('  node jira-workflow-demo.mjs batch-start SCRUM-55 SCRUM-56 SCRUM-57');
    console.log('  node jira-workflow-demo.mjs batch-complete SCRUM-55 SCRUM-56 SCRUM-57');
    console.log('  node jira-workflow-demo.mjs status SCRUM-55');
    return;
  }

  const command = args[0];
  const issueKeys = args.slice(1);

  switch (command) {
    case 'demo':
      await demoWorkflow(issueKeys[0] || 'SCRUM-55');
      break;

    case 'start':
      await startWork(issueKeys[0]);
      break;

    case 'complete':
      await completeWork(issueKeys[0]);
      break;

    case 'reopen':
      await reopenWork(issueKeys[0]);
      break;

    case 'batch-start':
      await batchWorkflow(issueKeys, 'start');
      break;

    case 'batch-complete':
      await batchWorkflow(issueKeys, 'complete');
      break;

    case 'status':
      const status = await getIssueStatus(issueKeys[0]);
      console.log(`${issueKeys[0]}: ${status}`);
      break;

    default:
      console.log(`Unknown command: ${command}`);
  }
}

main().catch(console.error);
